### PR TITLE
Replace AnchorJS with a Hugo render hook

### PR DESCRIPTION
**Pull request by XhmikosR**
TODO:

- [x] move the levels to config.yml
- [x] a11y
- [x] maybe use `::after`?
- [x] add padding?
- [x] fix `:focus` if possible
- [x] test that there are no regressions

Idea taken from https://github.com/gohugoio/gohugoioTheme/pull/146

**Pros**:

- No JS
- ...

**Cons**:

- Works only with Markdown headings, but on the other hand our ToC only works with Markdown already
- ...

---

Preview: <https://deploy-preview-32953--twbs-bootstrap.netlify.app/docs/5.1/getting-started/introduction/>

### CODE REVIEW

**Comment by XhmikosR** [comment]
@mdo @ffoodd any thoughts?

**Comment by coliff** [comment]
heya - FYI- I used render hook code similar to this for a docs project and Algolia displayed the anchor `#` in search results window after the titles which wasn't ideal.
We changed it so that the `#` was displayed using `::after` like you noted - so would definitely recommend you do that for a nicer Algolia search UX.

**Comment by XhmikosR** [comment]
@ffoodd could you help me out with this please? ❤️

**Comment by ffoodd** [comment]
I'll try to 👌

File: site/assets/scss/_anchor.scss [hunk 1]
```diff
@@ -1,11 +1,16 @@
-.anchorjs-link {
+.anchor-link {
+  padding-left: .175rem;
```
Comment (ffoodd) [inline site/assets/scss/_anchor.scss:2]:
```suggestion
  padding: 0 .175rem;
```
Makes the focus indicator better balanced.

File: site/assets/scss/_anchor.scss [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
   font-weight: 400;
   color: rgba($link-color, .5);
-  @include transition(color .15s ease-in-out);
+  text-decoration: none;
+  opacity: 0;
+  @include transition(color .15s ease-in-out, opacity .15s ease-in-out);
 
+  :hover > &,
```
Comment (ffoodd) [inline site/assets/scss/_anchor.scss:9]: This is an interesting addition. We may also make the whole line clickable using our `.streched-link` helper—not sure it'd be better though, mostly a UX consideration.

File: site/layouts/_default/_markup/render-heading.html [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<h{{ .Level }} id="{{ .Anchor | safeURL }}">{{ .Text | safeHTML }}
+{{- if and (ge .Level 2) (le .Level 5) }}{{" " -}}
+<a class="anchor-link" href="#{{ .Anchor | safeURL }}" aria-label="Anchor">#</a>
```
Comment (ffoodd) [inline site/layouts/_default/_markup/render-heading.html:3]: `aria-label` value is questionable. @patrickhlauke Do we consider the heading context is enough, or should we try to complete the label using heading's content?

**Review by ffoodd — COMMENTED** [review]
Seems already good to me!

What `:focus` fix did you have in mind? It's working fine IMHO.
There's no need to use a pseudo-element to handle `#` here because `aria-label` erases text content from link's accessible name.

So yeah, that's really good I think :)

**Comment by ffoodd** [comment]
Regarding this being markdown headings only: since this is outputs a very simple markup, we should be able to add anchors manually on markup headings, shouldn't we? It may even be a simple shortcode to help computing the markup.

File: site/layouts/_default/_markup/render-heading.html (hunk 3 above)
Comment (patrickhlauke) [reply site/layouts/_default/_markup/render-heading.html:3]: yeah i'd probably go for something a bit more descriptive, like having `aria-label="Link to this section: {{ .Text | safeHTML }}"` perhaps. yes, admittedly a bit wordy, but it front-loads what the meaning of the link is and AT users can always choose not to listen to the entire thing

Comment (XhmikosR) [reply site/layouts/_default/_markup/render-heading.html:3]: Note that I used what anchor.js is already doing. I don't have a preference about this, whatever you guys think is better.

**Comment by XhmikosR** [comment]
> Seems already good to me!
> 
> What `:focus` fix did you have in mind? It's working fine IMHO. There's no need to use a pseudo-element to handle `#` here because `aria-label` erases text content from link's accessible name.
> 
> So yeah, that's really good I think :)

See @coliff's [comment](https://github.com/twbs/bootstrap/pull/32953#issuecomment-902790544) above.

Regarding `:focus`, I thought that we showed the anchor on focus but apparently we don't.

File: site/assets/scss/_anchor.scss (hunk 2 above)
Comment (XhmikosR) [reply site/assets/scss/_anchor.scss:9]: I just tried to achieve the same behavior as before. Feel free to experiment with your suggestion if you think is better :)

**Review by ffoodd — APPROVED** [review]
Made a few tweaks to polish this: improved `aria-label` after discussion, and (opinionated) made the anchor link visible when anchor is active (with `:target`).

Might help recalling the current URL targets an anchor, which I often forget about :D

Edit: also moved the `#` to a pseudo-element to follow @coliff 's recommendation.
